src/envoy/http/authn/authn_utils.cc

```diff
@@ -0,0 +1,123 @@
+/* Copyright 2018 Istio Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "authn_utils.h"
+#include "common/json/json_loader.h"
+#include "src/envoy/http/jwt_auth/jwt.h"
+
+namespace Envoy {
+namespace Http {
+namespace Istio {
+namespace AuthN {
+namespace {
+class JsonIterCallback : public Logger::Loggable<Logger::Id::filter> {
+ public:
+  // The json iterator callback function
+  static bool JsonIteratorCallback(const std::string&,
+                                   const Envoy::Json::Object&,
+                                   istio::authn::JwtPayload* payload);
+};
+
+bool JsonIterCallback::JsonIteratorCallback(const std::string& key,
+                                            const Envoy::Json::Object& obj,
+                                            istio::authn::JwtPayload* payload) {
+  ENVOY_LOG(debug, "{}: key is {}", __FUNCTION__, key);
+  ::google::protobuf::Map< ::std::string, ::std::string>* claims =
+      payload->mutable_claims();
+  try {
+    if (key == "aud") {
```

Comment:
Sorry I miss this usecase. We do want to keep all original claims in 'claims' (may be as JsonString to be compatible with all types). Then for audiences and presenters (and user), parse the object again to construct the payload properly.

You can make change that in the following up PR.

Reply:
Will change that in the following up PR.

Reply:
Done, in the PR #1284.